### PR TITLE
ENH: Move tools out of Main Window and Application

### PR DESCRIPTION
**Pull request by hhslepicka**
This PR follows the idea of removing the coupling between PyDM Widgets and functionality and the PyDMApplication and PyDM Main Window components.

### CODE REVIEW

File: pydm/tools/__init__.py [hunk 1]
```diff
@@ -1 +1,153 @@
...
+            base_dir, _, _ = path_info(tool)
+            sys.path.append(base_dir)
+            temp_name = str(uuid.uuid4())
+
+            module = imp.load_source(temp_name, tool)
+            classes = [obj for _, obj in inspect.getmembers(module)
+                       if inspect.isclass(obj) and issubclass(obj,
+                                                              ExternalTool) and obj != ExternalTool]
+            if len(classes) == 0:
+                raise ValueError(
```
Comment (teddyrendahl) [inline pydm/tools/__init__.py]: Do you really want to `raise` here? It will crash your whole program...

File: pydm/tools/__init__.py [hunk 2]
```diff
@@ -1 +1,153 @@
...
+
+def load_external_tools():
+    """
+    Loads all the external tools available at the given
+    `PYDM_TOOLS_PATH` environment variable and subfolders that
+    follows the `*_tool.py` and have classes that inherits from
+    the `pydm.tools.ExternalTool` class.
+    """
+    if not ext_tools:
+        logger.debug("*" * 80)
```
Comment (teddyrendahl) [inline pydm/tools/__init__.py]: Can we get rid of the banner?

File: pydm/tools/__init__.py (hunk 1 above)
Comment (teddyrendahl) [reply pydm/tools/__init__.py]: Nevermind, saw the `try` except above

File: pydm/tools/__init__.py [hunk 3]
```diff
@@ -1 +1,153 @@
...
+    Install an External Tool at the PyDMApplication and add it to the
+    main window Tools menu.
+
+    Parameters
+    ----------
+    tool : str or pydm.tools.ExternalTool
+        The full path to a file containing a ExternalTool definition
+        or an Instance of an ExternalTool.
+    """
+    def reorder_tools_dict():
```
Comment (teddyrendahl) [inline pydm/tools/__init__.py]: Why make this a function of its own?

**Comment by hhslepicka** [comment]
@teddyrendahl this is good for another round of review.

File: pydm/tools/__init__.py (hunk 3 above)
Comment (hhslepicka) [reply pydm/tools/__init__.py]: Changed at the latest commit.

File: pydm/widgets/base.py [hunk 4]
```diff
@@ -273,8 +274,7 @@ def generate_context_menu(self):
             menu = QMenu(parent=self)
 
         kwargs = {'channels': self.channels_for_tools(), 'sender': self}
-        if hasattr(self.app, 'assemble_tools_menu'):
-            self.app.assemble_tools_menu(menu, widget_only=True, **kwargs)
+        tools.assemble_tools_menu(menu, widget_only=True, **kwargs)
```
Comment (mattgibbs) [inline pydm/widgets/base.py]: I can’t test for myself right now - does this screw up the Designer right-click menu?

**Comment by hhslepicka** [comment]
Even if designer uses it, the code for the tools append to the menu and not
destroy it. I will test it tomorrow and get back to you on that.
I can wrap it with a is_designer call before setting the widget menu.

**Pull request merged**